include/picongpu/plugins/saxs/Saxs.hpp

```diff
@@ -0,0 +1,815 @@
+                dataWriter = pmacc::memory::makeUnique< SaxsWriter< float_X > >(
+                    pluginPrefix + "Output",
+                    fileExtension,
+                    "saxsOutput",
+                    outputLayout,
+                    compressionMethod,
+                    precisionCast< uint64_t >( numVectors ),
+                    q_step,
+                    amplitudeUnit
+                );
+                // Set the output period.
+                outputPeriod = pluginSystem::toTimeSlice( outputPeriod_s );
+            }
+        }
+
+
+        void pluginUnload( ) override
+        {
+            if ( !notifyPeriod.empty( ) )
+            {
+                CUDA_CHECK( cudaGetLastError( ) );
+            }
+        }
+
+
+        //! Collect amplitude data from each CPU on the master node.
+        void collectIntensityOnMaster( )
+        {
+            auto const totalNumVectors = numVectors.productOfComponents( );
+
+            amplitude->deviceToHost( );
+            __getTransactionEvent( ).waitForFinished( );
+
+            reduce(
+                nvidia::functors::Add( ),
+                amplitudeMaster.data( ),
+                amplitude->getHostBuffer( ).getBasePointer( ),
+                bufferSize_i,
+                mpi::reduceMethods::Reduce( )
+            );
+
+        }
+
+
+        //! Calculates the offset to the the currently processed output chunk.
+        HINLINE uint32_t calcOffset( uint32_t const & step )
+        {
+            uint32_t totalRotations = accumulatedRotations + step;
+            uint32_t chunk = mpiRank + countRanks;
+            chunk = ( (chunk - ( totalRotations % countRanks ) ) %
+                       countRanks );
+            return chunk * bufferSize_i;
+        }
+
+
+        //! Checks if this node hast the last output part.
+        HINLINE bool hasLastChunk ( uint32_t const & step )
+        {
+            uint32_t totalRotations = accumulatedRotations + step;
+            return mpiRank == ( countRanks - 1 + totalRotations ) % countRanks;
+        }
+
+
+        //! Writes amplitude data to disk.
+        HINLINE void writeOutput( )
+        {
+            if ( outputLayout == OutputMemoryLayout::SplitOverNodes )
+            {
+                amplitude->deviceToHost( );
+                __getTransactionEvent( ).waitForFinished( );
+                realPart = extractReal( amplitude->getHostBuffer( ) );
+                imgPart = extractImag( amplitude->getHostBuffer( ) );
+
+                uint64_t offset = precisionCast< uint64_t >( calcOffset(
+                    countRanks - 1 ) );
+                uint64_t extent;
+                if ( hasLastChunk( countRanks - 1 ) ) extent = resOfVectors;
+                else extent =  bufferSize_i;
+                ( * dataWriter )(
+                    currentStep,
+                    std::move( extent ),
+                    std::move( offset ),
```

Comment:
Are these needed? Should make no difference whether the sizes are xvalues.

Reply:
proppably not :see_no_evil: